scripts/metric.py

```diff
@@ -1,68 +1,154 @@
+    python metric.py actual_column predicted_col \\
+        --prediction predictions.csv \\
+        --metric metrics.json \\
+        --selected-metrics spearman
+    ```
+
+Example metric JSON output:
+    ```json
+    {
+        "Average Spearman": 0.8542
+    }
+    ```
+
+Adding Custom Metrics:
+    To add a new metric, define a function following this pattern:
+    ```python
+    def metric_<name>(actual_values: list[float], predicted_values: list[float]) -> tuple[str, float]:
+        result = custom_calculation(actual_values, predicted_values)
+        return ("Metric Display Name", result)
+    ```
 
 Functions:
-    calc: Calculate and save performance metrics from prediction output files
+    metric_spearman: Calculate Spearman rank correlation coefficient
+    calculate_selected_metrics: Dynamically discover and execute selected metric functions
+    evaluate: Calculate and save performance metrics from prediction CSV to JSON format
+    main: Command-line interface for metric calculation
 """
 
+import inspect
+import sys
 import argparse
+import json
 from pathlib import Path
 
 import polars as pl
-from pycm import ConfusionMatrix
+from scipy.stats import spearmanr
+
+
+def metric_spearman(
+    actual_values: list[float],
+    predicted_values: list[float],
+) -> tuple[str, float]:
+    spearman_corr, _ = spearmanr(actual_values, predicted_values)
+    return ("Average Spearman", spearman_corr)
+
+def calculate_selected_metrics(
+    actual_values: list[float],
+    predicted_values: list[float],
+    selected_metrics: list[str],
+) -> dict[str, float]:
+    """Calculate selected custom metrics using auto-discovered metric functions.
+
+    This function dynamically discovers all functions with the 'metric_' prefix in the
+    current module and calls the requested ones. To add a new custom metric, define a
+    function with the naming pattern 'metric_<name>' that takes actual_values and
+    predicted_values as parameters and returns a tuple of (metric_name, metric_value).
+
+    Example:
+        def metric_custom(actual_values, predicted_values):
+            result = custom_calculation(actual_values, predicted_values)
+            return ("Custom Metric Name", result)
 
+    Args:
+        actual_values: List of actual/ground truth values
+        predicted_values: List of predicted values
+        selected_metrics: List of metric names to calculate (e.g., ["spearman"])
+                         Names should match the function suffix after 'metric_'
 
-def calc(
-    output: Path, metric: Path, actual_vector_col: str, predict_vector_col: str
+    Returns:
+        Dictionary mapping metric names to their computed values
+    """
+    current_module = sys.modules[__name__]
+    metric_functions = {}
+
+    for name, obj in inspect.getmembers(current_module, inspect.isfunction):
+        if name.startswith("metric_"):
+            metric_name = name.replace("metric_", "", 1)
+            metric_functions[metric_name] = obj
```

Comment:
I would make this a global dictionairy as it is a globally constant, but can also be recomputed every time this function is called